New challenges for Rc, Arc, and related Weak implementations (#367)
model-checking/verify-rust-std: doc/src/challenges/0026-rc.md

# Challenge 26: Verify reference-counted Cell implementation

- **Status:** Open
- **Solution:** *Option field to point to the PR that solved this challenge.*
- **Tracking Issue:** [#382](https://github.com/model-checking/verify-rust-std/issues/382)
- **Start date:** *2025/06/01*
- **End date:** *2025/12/31*
- **Reward:** *5,000 USD*

-------------------

## Goal

> The goal of this challenge is to verify [Rc](https://github.com/rust-lang/rust/blob/master/library/alloc/src/rc.rs) (meaning "Reference counted") and its related [Weak](https://github.com/rust-lang/rust/blob/master/library/alloc/src/rc.rs) implementation. Rc is the library-provided building block that enables safe multiple ownership of data through reference counting for single-threaded cases, as opposed to the usual ownership types used by Rust. A related challenge verifies the Arc implementation, which is atomic multi-threaded.

Reviewer: ```suggestion
The goal of this challenge is to verify [Rc](https://doc.rust-lang.org/std/rc/struct.Rc.html) (meaning "Reference counted") and its related [Weak](https://doc.rust-lang.org/std/rc/struct.Weak.html) implementation. Rc is the library-provided building block that enables safe multiple ownership of data through reference counting for single-threaded cases, as opposed to the usual ownership types used by Rust. A related challenge verifies the Arc implementation, which is atomic multi-threaded.
```

## Motivation

The Rc (for single-threaded code) and Arc (atomic multi-threaded) cell types are widely used in Rust programs to enable shared ownership of data through reference counting. Since shared ownership is generally not permitted by Rust's type system, these implementations use unsafe code to bypass Rust's usual compile-time checks. Verifying the Rust standard library thus fundamentally requires verification of these types.

## Description

This challenge verifies a number of Rc methods that encapsulate unsafety, as well as providing contracts for unsafe methods that impose safety conditions on their callers for correct use.

A key part of the Rc implementation is the Weak struct, which allows keeping a temporary reference to an Rc without creating circular references and without protecting the inner value from being dropped.

### Assumptions

Some properties needed for safety are beyond the ability of the Rust type system to express. This is true for all challenges, but we point out some of the properties that are relevant for this challenge.

> * It may be possible to use a new construct analogous to the [can_dereference API](https://model-checking.github.io/kani/crates/doc/kani/mem/fn.can_dereference.html). Our new construct would track that a pointer indeed originates from `into_raw`.

Reviewer: So what is the assumption here? This seems more like an observation. Is it incumbent on the person solving the challenge to propose and use such an API in their verification? Or is your point that for `into_raw` functions in `Rc` and `Weak`, they don't need to verify that the pointer returned originates from the given `Rc/Weak`? I think the latter, but I'd prefer to rephrase this bullet in terms of the assumption being made.

Author: Upon looking into it a bit more, we figured we might be able to use a technique similar to how `uninit-checks` works in order to track if a pointer originated from `into_raw`, so we removed this from the assumption section (i.e., we're leaving this as necessary for the completion of the challenge)

> * It is unclear how to show the reference count is greater than 0 when it is being decremented; the proposed `linked_list` [challenge](0005-linked-list.md) solution does not appear to check list length before performing operations either.

Reviewer: ```suggestion
* You are not required to check that the reference count is greater than 0 when it is being decremented
```
The reference to the linked list solution will become outdated if/when the solution moves from proposed to accepted, and I don't think it's necessary to make your point.
